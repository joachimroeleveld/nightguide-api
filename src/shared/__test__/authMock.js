const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
