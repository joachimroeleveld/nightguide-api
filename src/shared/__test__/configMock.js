const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
