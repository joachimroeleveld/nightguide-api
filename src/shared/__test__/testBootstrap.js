const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
