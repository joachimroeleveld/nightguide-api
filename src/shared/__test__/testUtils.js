const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
