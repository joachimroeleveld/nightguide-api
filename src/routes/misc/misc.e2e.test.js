require('../../shared/__test__/testBootstrap');

const request = require('supertest');
const sinon = require('sinon');

const { clearDb } = require('../../shared/__test__/testUtils');
const { validator } = require('../../shared/openapi');
const mailService = require('../../shared/services/mail');

const sandbox = sinon.createSandbox();

describe('users e2e', () => {
  afterEach(async () => {
    sandbox.restore();
    await clearDb();
  });

  describe('POST /misc/user-feedback', () => {
    const validateResponse = validator.validateResponse(
      'post',
      '/misc/user-feedback'
    );

    it('happy path', async () => {
      sandbox.stub(mailService, 'sendBasicEmail').resolves();

      const res = await request(global.app)
        .post('/misc/user-feedback')
        .send({
          message: 'test message',
        });

      expect(res.status).toEqual(200);
      expect(mailService.sendBasicEmail.getCall(0).args).toMatchSnapshot();
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /misc/city-config', () => {
    it('happy path', async () => {
      const res = await request(global.app).get('/misc/city-config');

      expect(res.status).toEqual(200);
      expect(res.body).toMatchSnapshot();
    });
  });
});
