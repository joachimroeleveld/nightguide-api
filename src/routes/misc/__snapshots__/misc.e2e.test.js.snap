// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`misc e2e POST /misc/user-feedback happy path 1`] = `
Array [
  "feedback@nightguide.app",
  "User app feedback",
  "
<strong>User alice@rogers.nl sent you message:</strong><br/>
<br/>
<p style=\\"font-style:italic;\\">test message</p>    
    ",
  Object {
    "replyTo": "alice@rogers.nl",
  },
]
`;
